docs/cqrs/pipeline/avoid_commiting_transactions_in_handlers.md

```diff
@@ -0,0 +1,137 @@
+
+```csharp
+public class CommitDatabaseTransactionMiddleware<TDbContext>
+    where TDbContext : DbContext
+{
+    private readonly RequestDelegate next;
+
+    public CommitDatabaseTransactionMiddleware(RequestDelegate next)
+    {
+        this.next = next;
+    }
+
+    public async Task InvokeAsync(HttpContext httpContext, TDbContext dbContext)
+    {
+        await next(httpContext);
+        await dbContext.SaveChangesAsync(httpContext.RequestAborted);
+    }
+}
+```
+
+Take, for instance, an example where we wish to send an email to an employee upon the occurrence of the `EmployeeAssignedToAssignment` event:
+
+```csharp
+public class Project : IAggregateRoot<ProjectId>
+{
+    // . . .
+
+    public void AssignEmployeeToAssignment(
+        AssignmentId assignmentId,
+        EmployeeId employeeId)
+    {
+        assignments.Single(t => t.Id == assignmentId)
+            .AssignEmployee(employeeId);
+
+        DomainEvents.Raise(new EmployeeAssignedToAssignment(
+            AssignmentId assignmentId,
+            EmployeeId employeeId));
+    }
+
+    // . . .
+}
+```
+
+Now, let's imagine using this method in a [command] handler:
+
+```csharp
+public class AssignEmployeeToAssignmentCH
+    : ICommandHandler<AssignEmployeeToAssignment>
+{
+    private readonly IRepository<Project, ProjectId> projects;
+    private readonly CoreDbContext dbContext;
+
+    public AssignEmployeeToAssignmentCH(
+        IRepository<Project, ProjectId> projects,
+        CoreDbContext dbContext)
+    {
+        this.projects = projects;
+        this.dbContext = dbContext;
+    }
+
+    public Task ExecuteAsync(HttpContext context, UpdateProjectName command)
+    {
+        var project = await projects.FindAndEnsureExistsAsync(
+            ProjectId.Parse(command.ProjectId),
+            context.RequestAborted);
+
+        project.AssignEmployeeToAssignment(
+            AssignmentId.Parse(command.AssignmentId),
+            EmployeeId.Parse(command.EmployeeId));
+
+        projects.Update(project);
+
+        // Directly committing the transaction in the command handler,
+        // which should be avoided.
+        await dbContext.SaveChangesAsync(context.RequestAborted)
+    }
+}
+```
+
+In our pipeline configuration, [events] are published after the [command] handler is executed. This implies that changes on `CoreDbContext` will be committed before [events] are published. In LeanCode Corelibrary, we utilize [MassTransit] for message broker interaction, employing the [transactional outbox](https://masstransit.io/documentation/patterns/transactional-outbox) concept. This ensures that messages are committed to the database before being available to message brokers after publication and the invocation of the `SaveChangesAsync` method on CoreDbContext.
```

Comment:
```suggestion
In our pipeline configuration, [events] are published after the [command] handler is executed. This implies that changes on `CoreDbContext` will be committed before [events] are published. In LeanCode CoreLibrary, we utilize [MassTransit] for message broker interaction, employing the [transactional outbox](https://masstransit.io/documentation/patterns/transactional-outbox) concept. This ensures that messages are committed to the database before being available to message brokers after publication and the invocation of the `SaveChangesAsync` method on CoreDbContext.
```

Also, the idea behind transactional outbox is not to wait for changes to the data before publishing, but to ensure the messages will be published **if and only if** the changes were made, so that the messages are neither lost, nor published even if save failed.